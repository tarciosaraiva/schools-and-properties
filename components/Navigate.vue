<template>
  <form class="navigate-container" @submit.prevent>
    <label for="geo_go_to" class="visually-hidden">Navigate to</label>
    <div class="d-flex align-items-center search-box">
      <font-awesome-icon icon="search-location" style="left: 0.75rem; position: absolute; z-index: 100" />
      <input
        id="geo_go_to"
        v-model="search"
        placeholder="Enter address, suburb, region or area to navigate"
        type="text"
        name="geo_go_to"
        autocomplete="false"
        @keyup="handleKeyUp"
        @keyup.down="onArrowDown"
        @keyup.up="onArrowUp"
        @keyup.enter="onEnter"
      />
    </div>
    <ul v-show="isOpen" class="geo-suggestions">
      <li v-for="(sug, i) in suggestions" :key="i" :class="{ 'is-active': i === arrowCounter }">
        <a href="#"
          class="suggestion"
          @click.stop.prevent="navigateTo(sug)">
          {{ sug.name }}
        </a>
      </li>
    </ul>
  </form>
</template>

<script lang="ts">
import Vue from 'vue'
import debounce from 'lodash.debounce'

export default Vue.extend({
  props: {
    suggestions: {
      type: Array,
      default: () => [] as any[]
    },
    geocodeFn: {
      type: Function,
      default: () => () => {}
    }
  },
  data() {
    return {
      search: '',
      isOpen: false,
      arrowCounter: -1
    }
  },

  watch: {
    suggestions (newSuggestions, _) {
      if (newSuggestions.length) {
        this.isOpen = true
      }
    }
  },
  created () {
    this.geolocate = debounce(this.geolocate, 300, { leading: false, trailing: true })
  },
  mounted() {
    document.addEventListener('click', this.handleClickOutside)
  },
  destroyed() {
    document.removeEventListener('click', this.handleClickOutside)
  },
  methods: {
    handleClickOutside(event: any) {
      if (!this.$el.contains(event.target)) {
        this.isOpen = false;
        this.arrowCounter = -1;
      }
    },
    handleFocus() {
      if (this.suggestions.length) {
        this.isOpen = true
      }
    },
    handleKeyUp() {
      this.geolocate()
    },
    geolocate () {
      this.geocodeFn(this.search)
    },
    navigateTo (location: any) {
      this.isOpen = false
      this.search = location.name
      this.$emit('navigate', location.center, location.placeType)
    },
    onArrowDown() {
      if (this.arrowCounter < this.suggestions.length) {
        this.arrowCounter = this.arrowCounter + 1;
      }
    },
    onArrowUp() {
      if (this.arrowCounter > 0) {
        this.arrowCounter = this.arrowCounter - 1;
      }
    },
    onEnter() {
      this.navigateTo(this.suggestions[this.arrowCounter])
      this.arrowCounter = -1;
    }
  }
})
</script>
