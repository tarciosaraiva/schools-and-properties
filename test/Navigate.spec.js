import { mount, createLocalVue } from '@vue/test-utils'
import { library } from '@fortawesome/fontawesome-svg-core'
import { fas } from '@fortawesome/free-solid-svg-icons'
import { FontAwesomeIcon } from '@fortawesome/vue-fontawesome'

import Navigate from '~/components/Navigate.vue'

library.add(fas)

const localVue = createLocalVue()
localVue.component('FontAwesomeIcon', FontAwesomeIcon)

const executeSearch = async (wrapper) => {
  const input = wrapper.find('#geo_go_to')
  input.setValue('my suburb')
  input.trigger('keyup')

  await wrapper.setProps({
    suggestions: [
      { name: 'my suburb', center: [1, 1], placeType: 'suburb' },
      { name: 'my other suburb', center: [2, 2], placeType: 'street' }
    ]
  })

  expect(wrapper.vm.search).toBe('my suburb')
  expect(wrapper.vm.isOpen).toBeTruthy()
}

describe('Navigate', () => {
  let wrapper
  let propsData

  beforeEach(() => {
    jest.useFakeTimers('modern')
    propsData = { searchCallback: jest.fn() }
    wrapper = mount(Navigate, { propsData, localVue })
  })

  afterEach(() => {
    jest.useRealTimers()
  })

  test('is a Vue instance', () => {
    expect(wrapper.vm).toBeTruthy()
  })

  test('renders correctly without suggestions', () => {
    expect(wrapper).toMatchSnapshot()
  })

  test('should trigger searchCallback when debounced', () => {
    const input = wrapper.find('#geo_go_to')
    input.setValue('my suburb')
    input.trigger('keyup')

    // debounce function should not have been triggered yet
    jest.advanceTimersByTime(200)
    expect(propsData.searchCallback).not.toHaveBeenCalled()

    // when 400ms have passed debounce function should have been triggered
    // and geolocate function is called and returned by now
    jest.advanceTimersByTime(200)
    expect(propsData.searchCallback).toHaveBeenCalledWith('my suburb')
  })

  test('renders correctly with suggestions', async () => {
    await executeSearch(wrapper)
    expect(wrapper).toMatchSnapshot()
  })

  test('should have loaded 2 suggestions', async () => {
    await executeSearch(wrapper)
    const suggestions = wrapper.findAll('.navigate-container .geo-suggestions .suggestion')
    expect(suggestions).toHaveLength(2)
  })

  test('should emit "navigate" event a suggestion is clicked', async () => {
    await executeSearch(wrapper)
    const suggestions = wrapper.findAll('.geo-suggestions .suggestion')
    suggestions.at(0).trigger('click')

    expect(wrapper.emitted().navigate).toBeTruthy()
  })

  test('should navigate with down arrow', async () => {
    await executeSearch(wrapper)
    const input = wrapper.find('#geo_go_to')
    await input.trigger('keyup.down')
    await input.trigger('keyup.down')
    const suggestion = wrapper.find('.navigate-container .geo-suggestions .is-active .suggestion')
    expect(suggestion.text()).toBe('my other suburb')
  })

  test('should navigate with up arrow', async () => {
    await executeSearch(wrapper)
    const input = wrapper.find('#geo_go_to')
    await input.trigger('keyup.down')
    await input.trigger('keyup.down')
    await input.trigger('keyup.up')
    const suggestion = wrapper.find('.navigate-container .geo-suggestions .is-active .suggestion')
    expect(suggestion.text()).toBe('my suburb')
  })

  test('should navigate with enter', async () => {
    await executeSearch(wrapper)
    const input = wrapper.find('#geo_go_to')
    await input.trigger('keyup.down')
    await input.trigger('keyup.down')
    await input.trigger('keyup.enter')

    expect(wrapper.emitted().navigate).toBeTruthy()
    expect(wrapper.vm.search).toBe('my other suburb')
    expect(wrapper.vm.isOpen).toBeFalsy()
  })
})
