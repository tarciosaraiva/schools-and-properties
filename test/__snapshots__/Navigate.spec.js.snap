// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navigate renders correctly with suggestions 1`] = `
<form class="navigate-container">
  <label class="visually-hidden" for="geo_go_to">Navigate to</label>
  <div class="d-flex align-items-center search-box">
    <svg aria-hidden="true" class="svg-inline--fa fa-search-location fa-w-16" data-icon="search-location" data-prefix="fas" focusable="false" role="img" style="left: 0.75rem; position: absolute; z-index: 100;" viewBox="0 0 512 512" xmlns="http://www.w3.org/2000/svg">
      <path class="" d="M505.04 442.66l-99.71-99.69c-4.5-4.5-10.6-7-17-7h-16.3c27.6-35.3 44-79.69 44-127.99C416.03 93.09 322.92 0 208.02 0S0 93.09 0 207.98s93.11 207.98 208.02 207.98c48.3 0 92.71-16.4 128.01-44v16.3c0 6.4 2.5 12.5 7 17l99.71 99.69c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.59.1-33.99zm-297.02-90.7c-79.54 0-144-64.34-144-143.98 0-79.53 64.35-143.98 144-143.98 79.54 0 144 64.34 144 143.98 0 79.53-64.35 143.98-144 143.98zm.02-239.96c-40.78 0-73.84 33.05-73.84 73.83 0 32.96 48.26 93.05 66.75 114.86a9.24 9.24 0 0 0 14.18 0c18.49-21.81 66.75-81.89 66.75-114.86 0-40.78-33.06-73.83-73.84-73.83zm0 96c-13.26 0-24-10.75-24-24 0-13.26 10.75-24 24-24s24 10.74 24 24c0 13.25-10.75 24-24 24z" fill="currentColor" />
    </svg>
    <input autocomplete="false" id="geo_go_to" name="geo_go_to" placeholder="Enter address, suburb, region or area to navigate" type="text">
  </div>
  <ul class="geo-suggestions" style>
    <li class>
      <a class="suggestion" href="#">
        my suburb
      </a>
    </li>
    <li class>
      <a class="suggestion" href="#">
        my other suburb
      </a>
    </li>
  </ul>
</form>
`;

exports[`Navigate renders correctly without suggestions 1`] = `
<form class="navigate-container">
  <label class="visually-hidden" for="geo_go_to">Navigate to</label>
  <div class="d-flex align-items-center search-box">
    <svg aria-hidden="true" class="svg-inline--fa fa-search-location fa-w-16" data-icon="search-location" data-prefix="fas" focusable="false" role="img" style="left: 0.75rem; position: absolute; z-index: 100;" viewBox="0 0 512 512" xmlns="http://www.w3.org/2000/svg">
      <path class="" d="M505.04 442.66l-99.71-99.69c-4.5-4.5-10.6-7-17-7h-16.3c27.6-35.3 44-79.69 44-127.99C416.03 93.09 322.92 0 208.02 0S0 93.09 0 207.98s93.11 207.98 208.02 207.98c48.3 0 92.71-16.4 128.01-44v16.3c0 6.4 2.5 12.5 7 17l99.71 99.69c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.59.1-33.99zm-297.02-90.7c-79.54 0-144-64.34-144-143.98 0-79.53 64.35-143.98 144-143.98 79.54 0 144 64.34 144 143.98 0 79.53-64.35 143.98-144 143.98zm.02-239.96c-40.78 0-73.84 33.05-73.84 73.83 0 32.96 48.26 93.05 66.75 114.86a9.24 9.24 0 0 0 14.18 0c18.49-21.81 66.75-81.89 66.75-114.86 0-40.78-33.06-73.83-73.84-73.83zm0 96c-13.26 0-24-10.75-24-24 0-13.26 10.75-24 24-24s24 10.74 24 24c0 13.25-10.75 24-24 24z" fill="currentColor" />
    </svg>
    <input autocomplete="false" id="geo_go_to" name="geo_go_to" placeholder="Enter address, suburb, region or area to navigate" type="text">
  </div>
  <ul class="geo-suggestions" style="display: none;"></ul>
</form>
`;
